### ISSUE TITLE
KCoreAddons is missing in the Packaged Building page

### ISSUE THREAD

**Issue by catdevnull** [telegramdesktop/tdesktop#25073]
KCoreAddons is missing in the page even though it's required since https://github.com/telegramdesktop/tdesktop/commit/e20840b4d4c828d930b55a93995ded9eb2784c2d

**Comment by ilya-fedin** [comment]
Yeah, I don't know what would be the best way to explain its building at its trickery to have the Qt 6 version along with the Qt 5 version. Would be really nice to hear anyone's practical experience on this. I see two options at least: telling it to install to a custom libdir (`-DCMAKE_INSTALL_LIBDIR`) or set the installation prefix to a subdirectory in building tree and build it along with tdesktop (`-DCMAKE_INSTALL_PREFIX`, `-DBUILD_SHARED_LIBS=OFF`), not sure what would be better. Either way it's possible to help cmake find the library with `-DKF5CoreAddons_DIR`.

**Comment by xvitaly** [comment]
KCoreAddons must be bundled, because packaged version built against Qt 5 in all distributions.

**Comment by ilya-fedin** [comment]
No, it definitely not 'must', there are other options, see my previous message

**Comment by xvitaly** [comment]
> No, it definitely not 'must', there are other options, see my previous message

Maybe it should be done automatically with ExternalProject_Add?

**Comment by ilya-fedin** [comment]
I'd like to avoid that due to two reasons:
1. It's not possible to use extra-cmake-modules without installation (and I don't like that hack I made the previous time for that, having extra-cmake-modules in Docker image is also not an option as KDE frameworks require ECM of the same version)
2. kcoreaddons requires qttools for build and I don't want to have it in final Docker image

**Comment by catdevnull** [comment]
Context: I'm the maintainer for telegram-desktop in Alpine Linux and am a Qt/CMake noob. I ended up making a [separate kcoreaddons-qt6 package](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/39164). It seems to have built fine although I haven't tested it. Is this wrong?

**Comment by ilya-fedin** [comment]
And it doesn't conflict with the qt5 one?
![изображение](https://user-images.githubusercontent.com/17829319/190873945-42399b4e-d093-41c0-8f73-49a8f76c528c.png)
The idea is to set e.g. `-DCMAKE_INSTALL_LIBDIR=lib/kf5qt6`, so it won't conflict with the qt5 build.
That's how e.g. ffmpeg 4.4 is built to be installed along with ffmpeg 5.0: https://github.com/archlinux/svntogit-packages/blob/930ca4335cc72b7badab2b278790d23ca42d7031/trunk/PKGBUILD#L118. KDE frameworks install include headers to KF6 folder when building with Qt 6, so it shouldn't conflict, but maybe you'd need to remove the content of share directory from the qt6 package and add suffix to binaries (e.g. -qt6) to avoid conflict. The content of datadir from the qt5 package could be separated to make it a dependency of both the qt5 and qt6 packages).

**Comment by ilya-fedin** [comment]
Then you set `-DKF5CoreAddons_DIR=/usr/lib/kf5qt6/lib/cmake/KF5CoreAddons` when building tdesktop and cmake will find the library

**Comment by xvitaly** [comment]
> I ended up making a [separate kcoreaddons-qt6 package](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/39164). It seems to have built fine although I haven't tested it. Is this wrong?

Not an option. It will conflict with kf5-kcoreaddons, which is unacceptable.

**Comment by catdevnull** [comment]
Yes, they do conflict, which is probably bad.

**Comment by xvitaly** [comment]
I see only one acceptable option - build KF6CoreAddons before the telegram-desktop in its build manifest as a static library and then use `-DKF5CoreAddons_DIR`.

**Comment by ilya-fedin** [comment]
It's the simplest option if you can't contact kcoreaddons maintainers, yeah

**Comment by xvitaly** [comment]
> It's the simplest option if you can't contact kcoreaddons maintainers, yeah

kcoreaddons won't built it against Qt 6, because all other KF5 packages built against Qt 5.

**Comment by ilya-fedin** [comment]
I don't see why not. It's a matter of need. If a package starts depending on Qt 6 version of the package, I don't see a reason to not to add it. Especially given that all its dependencies (qtbase, qttools, extra-cmake-modules) are in place

**Comment by catdevnull** [comment]
It looks to me that [kcoreaddons](https://github.com/KDE/kcoreaddons/search?p=1&q=qt6&type=commits) is at the very least being built with Qt6 in their CI.

**Comment by xvitaly** [comment]
> I don't see why not. It's a matter of need. If a package starts depending on Qt 6 version of the package, I don't see a reason to not to add it.

Because all other KDE frameworks, apps and KDE itself built against Qt 5.

**Comment by ilya-fedin** [comment]
> It looks to me that kcoreaddons is at the very least [being built with Qt6](https://github.com/KDE/kcoreaddons/search?p=1&q=qt6&type=commits) in their CI.

We're talking about distro maintainers :)
But I guess contacting upstream for more onvenient options is an option, too.

> Because all other KDE frameworks, apps and KDE itself built against Qt 5.

Well, I don't see any barrier here. Now you have a first Qt 6 one, the time has come.

**Comment by xvitaly** [comment]
> I don't see why not. It's a matter of need. If a package starts depending on Qt 6 version of the package, I don't see a reason to not to add it.

Major distribution won't do this in a foreseen future due to potential bugs until KDE upstream switches to Qt 6.

**Comment by mid-kid** [comment]
> contacting upstream for more convenient options is an option, too.

Yeah, installing in different paths isn't a huge deal, but it'd be nice to get a new pkgconfig filename for upstreams to build against either.

**Comment by ilya-fedin** [comment]
> Major distribution won't do this in a foreseen future due to potential bugs until KDE upstream switches to Qt 6.

I wouldn't speak on behalf of them on your place

**Comment by mid-kid** [comment]
> Major distribution won't do this in a foreseen future due to potential bugs until KDE upstream switches to Qt 6.

Again, you're missing the part where you can do this in a way that doesn't replace or conflict with the existing qt5-based framework libraries. If people start depending on them there will be no choice.

**Comment by xvitaly** [comment]
> Again, you're missing the part where you can do this in a way that doesn't replace or conflict with the existing qt5-based framework libraries. If people start depending on them there will be no choice.

Then all KF5 should be built against Qt 6. Mixing Qt 5 and Qt 6 in one project is a very bad idea.

**Comment by xvitaly** [comment]
Btw, KDE still doesn't support theming for Qt 6. All Qt 6 dialogs looks too ugly:

![image](https://user-images.githubusercontent.com/966711/190875550-0d756fe6-474f-42ad-aa10-10f7ebe7f0a2.png)

**Comment by mid-kid** [comment]
> Then all KF5 should be built against Qt 6. Mixing Qt 5 and Qt 6 in one project is a very bad idea.

Nobody is doing this. Having the libraries installed side-by-side doesn't mean one program will load both.

**Comment by mid-kid** [comment]
> KDE still doesn't support theming for Qt 6

Then build tdesktop for Qt 5 I guess.

**Comment by xvitaly** [comment]
> Then build tdesktop for Qt 5 I guess.

@ilya-fedin said that Qt 5 is not supported in telegram-desktop anymore.

**Comment by mid-kid** [comment]
I'm not ilya, but Qt 6 doesn't support windows 7, so as long as tdesktop supports win7 there'll be some sort of support for it.

Gentoo currently defaults to building with Qt 5 since Qt 6 was introduced to the distro very late and is still considered unstable, as well as reasons like the one you mentioned. Users have the option to unmask the flag and configure it to build with qt6, though, providing us with a way to test it for the time being. The only point of contention is wayland, where some integrations don't work with qt5 anymore, but I have no idea what that affects if anything. Anyway all I'm saying is using qt5 is still an option for the time being if you're having problems with qt6.

**Comment by ilya-fedin** [comment]
@xvitaly is right, Qt 5 is not supported on Linux in fact, it builds just because it builds, but no one tests it and it can break at any time.

**Comment by ilya-fedin** [comment]
> Then build tdesktop for Qt 5 I guess.

IMO that's not a reason to build with Qt 5. gtk applications have different dialogs yet people use them.

**Comment by ilya-fedin** [comment]
> Then all KF5 should be built against Qt 6. Mixing Qt 5 and Qt 6 in one project is a very bad idea.

I think you overcomplicate it, it shouldn't. And no one talks about mixing Qt 5 and Qt 6.

**Comment by ilya-fedin** [comment]
tdesktop has such a patch in Qt patchset for better Qt integration in the system: https://github.com/desktop-app/patches/blob/master/qtbase_6_4_0/0015-nicer-platformtheme-choosing.patch. It makes Qt always using portal dialogs for better integration and always fallback to gtk as last resort so icon theming works even in WM-only environments as well as in Plasma thanks to GTK compatibility layer.

**Comment by xvitaly** [comment]
> I think you overcomplicate it, it shouldn't. And no one talks about mixing Qt 5 and Qt 6.

Maybe. I need to discuss this issue with Fedora KDE team. As a workaround, I will build KF6CoreAddons as a static library before the telegram-desktop.

**Comment by svenstaro** [comment]
Hey, Arch maintainer here. Reading this thread, I think the best option for the time being would be to handle this like @xvitaly [suggested](https://github.com/telegramdesktop/tdesktop/issues/25073#issuecomment-1250048936) and also bundle it like KF5KWayland. This would be the smoothest by far and besides, you guys are already doing it so what's the harm in doing it for one more software?

**Comment by ilya-fedin** [comment]
Dependency on KF5Wayland was replaced multiple months ago with Qt6-based QtWaylandScanner support, all related cmake code was removed

**Comment by mid-kid** [comment]
> you guys are already doing it so what's the harm in doing it for one more software?

My main concern is having other software depend on KF5 built with Qt6, which would then need to bundle it as well. Bundling is the easy way out, but I'd prefer allowing parallel installation.

**Comment by svenstaro** [comment]
Telegram is the first software of this kind as far as I'm aware in Arch at least and also once Qt6 is stable with KF5, we'll likely switch the whole thing over anyway. There's no need to design a proper transition where all software can use KF5 with either Qt5 or Qt6 I think as experience shows it doesn't happen often at all. This is a case-by-case thing and we should pick the smoothest experience with that in mind.

**Comment by ilya-fedin** [comment]
If you convince KDE upstream to make qttools optional dependency and a way to set path to extra-cmake-modules without going through full cmake build&install command chain (so I don't have to add ExternalProject for extra-cmake-modules again), I'll switch kcoreaddons to ExternalProject

**Comment by antonio-rojas** [comment]
> If you convince KDE upstream to make qttools optional dependency and a way to set path to extra-cmake-modules without going through full cmake build&install command chain (so I don't have to add ExternalProject for extra-cmake-modules again), I'll switch kcoreaddons to ExternalProject

Building kcoreaddons from git does not require qttools since the git repo does not contain any translations. It is only needed if you build it from a release tarball.

**Comment by ilya-fedin** [comment]
> Building kcoreaddons from git does not require qttools since the git repo does not contain any translations. It is only needed if you build it from a release tarball.

If it was true I wouldn't talk about that, but cmake requires qttools as soon as it sees this line: https://github.com/KDE/kcoreaddons/blob/e48f96d13333639ef8ffb12450b00e0f1a5a6a91/CMakeLists.txt#L105

**Comment by antonio-rojas** [comment]
> > Building kcoreaddons from git does not require qttools since the git repo does not contain any translations. It is only needed if you build it from a release tarball.
> 
> If it was true I wouldn't talk about that, but cmake requires qttools as soon as it sees this line: https://github.com/KDE/kcoreaddons/blob/e48f96d13333639ef8ffb12450b00e0f1a5a6a91/CMakeLists.txt#L105

That is a no-op if there is no `po' dir, as it is the case in a git checkout.

**Comment by ilya-fedin** [comment]
I can't confirm this
```
#0 5.062 CMake Error at /usr/local/share/ECM/modules/ECMPoQmTools.cmake:151 (find_package):
#0 5.062   Found package configuration file:
#0 5.062 
#0 5.062     /usr/local/desktop-app/Qt-6.4.0/lib/cmake/Qt6/Qt6Config.cmake
#0 5.062 
#0 5.062   but it set Qt6_FOUND to FALSE so package "Qt6" is considered to be NOT
#0 5.062   FOUND.  Reason given by package:
#0 5.062 
#0 5.062   Failed to find required Qt component "LinguistTools".
#0 5.062 
#0 5.062   Expected Config file at
#0 5.062   "/usr/local/desktop-app/Qt-6.4.0/lib/cmake/Qt6LinguistTools/Qt6LinguistToolsConfig.cmake"
#0 5.062   does NOT exist
#0 5.062 
#0 5.062   
#0 5.062 
#0 5.062   Configuring with -DCMAKE_FIND_DEBUG_MODE=TRUE might reveal details why the
#0 5.062   package was not found.
#0 5.062 
#0 5.062   Configuring with -DQT_DEBUG_FIND_PACKAGE=ON will print the values of some
#0 5.062   of the path variables that find_package uses to try and find the package.
#0 5.062 
#0 5.062 Call Stack (most recent call first):
#0 5.062   /usr/local/share/ECM/modules/ECMPoQmTools.cmake:242 (ecm_process_po_files_as_qm)
#0 5.062   CMakeLists.txt:106 (ecm_install_po_files_as_qm)
```

**Comment by antonio-rojas** [comment]
Your backtrace points at line 242 of `ECMPoQmTools.cmake`. The code shows that this line can only be reached if there are `po` files. 

https://invent.kde.org/frameworks/extra-cmake-modules/-/blob/master/modules/ECMPoQmTools.cmake#L242

**Comment by ilya-fedin** [comment]
see this block, I guess it takes action
https://invent.kde.org/frameworks/extra-cmake-modules/-/blob/master/modules/ECMPoQmTools.cmake#L226-236

**Comment by antonio-rojas** [comment]
That block shows that the function returns if it can't find a `po` dir either in the source or in the cmake binary dir. So you must have it somewhere.

**Comment by ilya-fedin** [comment]
See the commit, ECM can download them automatically to the binary dir. I guess that's what it does.

**Comment by xvitaly** [comment]
> See the commit, ECM can download them automatically to the binary dir. I guess that's what it does actually.

Official distro builds has no network access btw.

**Comment by antonio-rojas** [comment]
> See the commit, ECM can download them automatically to the binary dir. I guess that's what it does actually.

It does if you set `KDE_L10N_AUTO_TRANSLATIONS` which is off by default

https://invent.kde.org/frameworks/extra-cmake-modules/-/blob/master/kde-modules/KDECMakeSettings.cmake#L331

**Comment by ilya-fedin** [comment]
It's still a problem for official builds...

**Comment by antonio-rojas** [comment]
> > It does if you set `KDE_L10N_AUTO_TRANSLATIONS` which is off by default
> 
> I don't set it, yet it I get the error. If you don't believe, try on your own. You can use tdesktop's Dockerfile and remove the `git submodule update --init --recursive qttools` line. That's enough for me to get the error.

OK, I see what's going on, you're checking out the git tag which has the translations injected. You should use the parent commit instead, `80cb2d7a47278d984c1bda38eae08ea113c8dff0`. Alternatively, remove the `po` dir manually after checkout.

**Comment by ilya-fedin** [comment]
Wow, so easy, thanks

**Comment by ilya-fedin** [comment]
So the only remaining problem is the necessity to "build" ECM...

**Comment by DeadMozay** [comment]
Adding KF5CoreAddons causes build error
`CMake Error: The INTERFACE_QT_MAJOR_VERSION property of "Qt5::Core" does not agree with the value of QT_MAJOR_VERSION already determined`

**Comment by ilya-fedin** [comment]
@DeadMozay you need to build a Qt6 version of KCoreAddons, all the thread is about that

**Comment by ilya-fedin** [comment]
I've achieved to make custom cmake build rules for kcoreaddons to build only the required parts. Unfortunately, there's still a blocker as we have to patch kcoreaddons as the vanilla version breaks pasting from gtk4 applications and there's no good way to patch in-tree copy of kcoreaddons (the only good way is to copy the tree to CMAKE_CURRENT_BINARY_DIR, but then there's no patch command on Windows).

If someone would convince KDE upstream to merge https://invent.kde.org/frameworks/kcoreaddons/-/merge_requests/258, then I'll add submodule pointing to that commit and push the updated cmake rules.

**Comment by mid-kid** [comment]
If this is an essential patch, please consider forking the library for the time being, and submoduling that.

**Comment by ilya-fedin** [comment]
I don't really want to do that :(

**Comment by xvitaly** [comment]
> I don't really want to do that :(

But you bundled kimageformats yesterday. One more, one less.

**Comment by ilya-fedin** [comment]
I don't want to fork, not to bundle

**Comment by xvitaly** [comment]
@ilya-fedin Why did you bundle kimageformats btw? It is already packaged and available in all GNU/Linux distributions.

**Comment by ilya-fedin** [comment]
It's not possible to use the system one when you have a custom Qt build

**Comment by mid-kid** [comment]
Wait, weren't you using qtimageformats at first? Why the switch?

**Comment by ilya-fedin** [comment]
It's not a switch, kimageformats contains additional codecs for modern formats like AVIF, HEIF and JPEG XL

**Comment by mid-kid** [comment]
I see. I happen to have it installed, but it's built without AVIF, HEIF and JPEG XL support and haven't noticed any issues yet. Will any of these formats be required for core functionality like stickers and gifs or is it just for received images? Just wondering whether I should make it a required dep like qtimageformats or a suggested dep.

**Comment by ilya-fedin** [comment]
> I happen to have it installed, but it's built without AVIF, HEIF and JPEG XL support and haven't noticed any issues yet.

It's for official builds, built whenever `NOT DESKTOP_APP_USE_PACKAGED OR DESKTOP_APP_USE_PACKAGED_LAZY` is true (just like the other third-party Qt  plugins used in official builds).

> Will any of these formats be required for core functionality like stickers and gifs or is it just for received images?

No, it's just to let people send those formats as photos (so tdesktop can detect and convert to jpeg them) or open with internal image viewer.

> Just wondering whether I should make it a required dep like qtimageformats or a suggested dep.

You may want to let people enable DESKTOP_APP_USE_PACKAGED_LAZY as that's probably the only way to let people use those modern formats with Qt 6 unless distros start provide kimageformats built with Qt 6.

**Comment by ilya-fedin** [comment]
Here's the full list of Qt plugins this macro builds: https://github.com/desktop-app/cmake_helpers/blob/master/external/qt/qt_static_plugins/qt_static_plugins.cpp#L57-L74

**Comment by mid-kid** [comment]
> You may want to let people enable DESKTOP_APP_USE_PACKAGED_LAZY

Considering how this has a tendency to cause automagic dependencies (I think, this flag has never been clear to me), I'd rather patch in a toggle for this specific instance of it, so we don't get taken by surprise with new deps in the future.

**Comment by ilya-fedin** [comment]
The only thing it does in addition to building the bundled Qt plugins is add QUIET to find_package calls of libraries that have bundled copies (GSL, tl-expected, range-v3)

> I think, this flag has never been clear to me

It's to consistently reproduce the set of Qt plugins in all the official builds. There's unfortunately no way to hardly link to a plugin without linking it statically and linking statically is not what everyone wants, so there's a switch...

**Comment by ilya-fedin** [comment]
I've pushed the cmake_helpers code to my fork in case anything happens to me
https://github.com/ilya-fedin/cmake_helpers/tree/bundled-kcoreaddons

The only things it also needs is to replace includes like <KUrlMimeData> to <kurlmimedata.h> and adding submodule

**Comment by catdevnull** [comment]
>The only things it also needs is to replace includes like <KUrlMimeData> to <kurlmimedata.h>

How would this be done?

**Comment by ilya-fedin** [comment]
> How would this be done?

Either I will PR the changes when (if) KDE accept my patch or someone else interested in case I disappear. Not sure I understood your question right though,

**Comment by catdevnull** [comment]
I don't understand how to replace the includes to make the bundled KCoreAddons work.

**Comment by ilya-fedin** [comment]
It's for anyone who would like to finish the work in case something happens to me. One would need to apply the cmake_helpers patch, add submodule and build tdesktop with fixing all the include errors like I described.

**Comment by catdevnull** [comment]
Okay, I think I understand. I tried applying the patch and adding the submodule but it won't build because it can't find the compiled version of the DBus portal thingy. I'll leave this be until this is properly supported. (https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/39162)

**Comment by ilya-fedin** [comment]
oops, I see the QT_VERSION_MAJOR condition is quite wrong looking at your MR
Regarding the crash, I can try to help if you will build with debug symbols and get a good stack trace

**Comment by catdevnull** [comment]
Sorry, the crash is when I was trying to build it as a separate package. It was loading the wrong version. Doing that would require more effort for seemingly no benefit compared to statically building (no package is using KCoreAddons in Qt6 for now), so I switched to that.

**Comment by ilya-fedin** [comment]
I've updated the commit with the right condition

**Comment by catdevnull** [comment]
This is the error I'm currently getting with the cmake patch: https://gitlab.alpinelinux.org/Nulo/aports/-/jobs/847932#L829

**Comment by ilya-fedin** [comment]
It seems to be some miracle, maybe some magic of cmake version used in the Dockerfile, or ninja, or whatever, but I pushed the fixed version

**Comment by ilya-fedin** [comment]
I've pushed again to fix build with dbus

**Comment by catdevnull** [comment]
Thanks a lot! It now fails to find KUrlMimeType: https://gitlab.alpinelinux.org/Nulo/aports/-/jobs/847983#L1367

**Comment by ilya-fedin** [comment]
Yeah, that's what I talked about, now you need to find and replace the `<KUrlMimeData>` includes (as well as __has_include checks)

**Comment by ilya-fedin** [comment]
In 4.2.0 it's only utils.h fortunately (it's more in dev :P)

**Comment by ilya-fedin** [comment]
`	sed -i 's/^#include <KUrlMimeData>/#include <kurlmimedata.h>/g' $(find . -iname '*.h' -or -iname '*.cpp')` won't replace _has_include checks, `	sed -i 's/^<KUrlMimeData>/<kurlmimedata.h>/g' $(find . -iname '*.h' -or -iname '*.cpp')` is the right way

**Comment by catdevnull** [comment]
I actually didn't, I copied it manually, I'm multitasking and I'm very bad at it.

**Comment by ilya-fedin** [comment]
@catdevnull looks like find and xargs can't communicate with null as separator, I tried your current command and it does nothing, but if I remove -print0 and -0 then it works

**Comment by catdevnull** [comment]
It's that it was only applying print0 in the .cpp case; I fixed it now.

**Comment by ilya-fedin** [comment]
ah

**Comment by catdevnull** [comment]
telegram-desktop 4.2.0 now builds and works in Alpine with Qt6 and bundled KCoreAddons patch from [ilya-fedin's branch](https://github.com/ilya-fedin/cmake_helpers/tree/bundled-kcoreaddons), see [the MR](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/39162) (and specifically [the prepare part](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/39162/diffs#744773445afaa91931415f9b05bc8cf85e9d6f4d_84_88) where the changes needed for bundled KCoreAddons to work are done)

**Comment by svenstaro** [comment]
We [fixed the package in Arch Linux](https://github.com/archlinux/svntogit-community/commit/5e3c500a47c9d8f5b2d00ec5b0b653a513c0f167) but a more convenient and less patchy solution would still be appreciated.
